### PR TITLE
[Feature] Segformer backbone re-implementation

### PR DESCRIPTION
**Pull request by sennnnn**
- [x]  Add backbone MixVisionTransformer;

~- [ ]  Add head SegFormerHead;~

~- [ ]  Add dataset transform pipeline AlignedResize;~

~- [ ]  Add some config for segformer;~

### CODE REVIEW

**Review by AncientRemember — APPROVED** [review]

**Comment by AncientRemember** [comment]
can you provide the config files

File: mmseg/models/backbones/mit.py [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+                 act_cfg=dict(type='GELU'),
+                 drop_rate=0.):
+        super().__init__()
+
+        in_channels = embed_dims
+        self.fc1 = Linear(in_channels, feedforward_channels)
+        self.dwconv = DWConv(feedforward_channels)
+        self.act = build_activation_layer(act_cfg)
+        self.fc2 = Linear(feedforward_channels, in_channels)
+        self.drop = nn.Dropout(drop_rate)
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:52]: We can replace FC with 1x1 conv, so we can avoid dimension transform.

File: mmseg/models/backbones/mit.py [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+    def forward(self, x, H, W):
+        B, _, C = x.shape
+        x = x.transpose(1, 2).reshape(B, C, H, W)
+        x = self.dwconv(x)
+        x = x.flatten(2).transpose(1, 2).contiguous()
+
+        return x
+
+
+class Mlp(BaseModule):
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:38]: Be careful that the variable or class name needs to be consistent with the paper.
Rename Mlp to MixFFN

File: mmseg/models/backbones/mit.py [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+    def forward(self, x, H, W):
+        x = self.fc1(x)
+        x = self.dwconv(x, H, W)
+        x = self.act(x)
+        x = self.fc2(x)
+        x = self.drop(x)
+        return x
+
+
+class Attention(BaseModule):
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:63]: Rename `Attention` to `EfficientMultiheadAttention`.
We can also inherit from `MultiheadAttention` in MMCV and pass different query, key, and value in forward phrase according to different spatial reduction.

File: mmseg/models/backbones/mit.py [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+        self.proj_drop = nn.Dropout(proj_drop)
+
+        self.sr_ratio = sr_ratio
+        if sr_ratio > 1:
+            self.sr = ConvModule(
+                in_channels=dim,
+                out_channels=dim,
+                kernel_size=sr_ratio,
+                stride=sr_ratio)
+            _, self.norm = build_norm_layer(norm_cfg, dim)
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:96]: If we inherit from `MultiheadAttention` in MMCV, we only need to add these lines.

File: mmseg/models/backbones/mit.py [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+        attn = self.attn_drop(attn)
+
+        x = (attn @ v).transpose(1, 2).reshape(B, N, C)
+        x = self.proj(x)
+        x = self.proj_drop(x)
+
+        return x
+
+
+class Block(BaseModule):
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:127]: Rename it.

File: mmseg/models/backbones/mit.py [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+        self.attn = Attention(
+            dim,
+            num_heads=num_heads,
+            qkv_bias=qkv_bias,
+            qk_scale=qk_scale,
+            attn_drop=attn_drop_rate,
+            proj_drop=drop_rate,
+            sr_ratio=sr_ratio)
+        # NOTE: drop path for stochastic depth, we shall see if this is better
+        # than dropout here
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:152]: Is this comment necessary?

File: mmseg/models/backbones/mit.py [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+                 embed_dim=768):
+        super().__init__()
+        img_size = to_2tuple(img_size)
+        patch_size = to_2tuple(patch_size)
+
+        self.img_size = img_size
+        self.patch_size = patch_size
+        num_rows, num_cols = img_size[0] // patch_size[0], img_size[
+            1] // patch_size[1]
+        self.num_patches = num_rows * num_cols
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:187]: Are these lines necessary?

File: mmseg/models/backbones/mit.py [hunk 8]
```diff
@@ -0,0 +1,420 @@
...
+        self.proj = nn.Conv2d(
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:188]: Use `ConvModule`

File: mmseg/models/backbones/mit.py [hunk 9]
```diff
@@ -0,0 +1,420 @@
...
+                 depths=[3, 4, 6, 3],
+                 sr_ratios=[8, 4, 2, 1],
+                 init_cfg=None,
+                 pretrained=None):
+        super().__init__()
+        self.init_cfg = init_cfg
+        self.pretrained = pretrained
+        self.depths = depths
+        # patch_embed
+        self.patch_embed1 = OverlapPatchEmbed(
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:236]: Use `ModuleList`

File: mmseg/models/backbones/mit.py [hunk 10]
```diff
@@ -0,0 +1,420 @@
...
+            patch_size=3,
+            stride=2,
+            in_chans=embed_dims[2],
+            embed_dim=embed_dims[3])
+
+        # transformer encoder
+        dpr = [
+            x.item() for x in torch.linspace(0, drop_path_rate, sum(depths))
+        ]  # stochastic depth decay rule
+        cur = 0
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:265]: Define `self.stages = ModuleList()` to contain self.block1,  self.block2,  self.block3, and  self.block4

File: mmseg/models/backbones/mit.py [hunk 11]
```diff
@@ -0,0 +1,420 @@
...
+        elif isinstance(self.pretrained, str):
+            logger = get_root_logger()
+            load_checkpoint(
+                self,
+                self.pretrained,
+                map_location='cpu',
+                strict=False,
+                logger=logger)
+
+    def reset_drop_path(self, drop_path_rate):
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:354]: Where is this used?

File: mmseg/models/backbones/mit.py [hunk 12]
```diff
@@ -0,0 +1,454 @@
...
+        else:
+            raise TypeError('pretrained must be a str or None')
+
+        self.out_indices = out_indices
+        self.pretrain_style = pretrain_style
+        self.pretrained = pretrained
+        self.init_cfg = init_cfg
+
+        patch_sizes = [7, 3, 3, 3]
+        strides = [4, 2, 2, 2]
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:367]: Make these arguments.

File: mmseg/models/backbones/mit.py [hunk 13]
```diff
@@ -0,0 +1,454 @@
...
+            norm_cfg=None,
+            act_cfg=None)
+
+    def forward(self, x):
+
+        x = self.conv(x)
+
+        return x
+
+
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:60]: `@FEEDFORWARD_NETWORK.register_module()`

File: mmseg/models/backbones/mit.py [hunk 14]
```diff
@@ -0,0 +1,454 @@
...
+class MixFFN(BaseModule):
+    """An implementation of MixFFN of Segformer.
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:62]: We may register this `FFN`  to `FEEDFORWARD_NETWORK` in MMCV

File: mmseg/models/backbones/mit.py [hunk 15]
```diff
@@ -0,0 +1,454 @@
...
+    def __init__(self,
+                 embed_dims,
+                 feedforward_channels,
+                 num_fcs=2,
+                 act_cfg=dict(type='GELU'),
+                 ffn_drop=0.,
+                 dropout_layer=None,
+                 add_identity=True,
+                 init_cfg=None):
+        super(MixFFN, self).__init__(init_cfg)
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:96]: If we consider the `3x3 conv` as Position Encoding (PE), we only need one and add an argument to indicate where to insert in `FCs`.

File: mmseg/models/backbones/mit.py [hunk 16]
```diff
@@ -0,0 +1,454 @@
...
+            act_cfg=act_cfg)
+
+    def forward(self, x, H, W):
+        x = self.attn(self.norm1(x), H, W, identity=x)
+        x = self.ffn(self.norm2(x), H, W, identity=x)
+        return x
+
+
+class OverlapPatchEmbed(BaseModule):
+    """Image to Patch Embedding."""
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:296]: We can move embedding layer to` mmseg/models/utils/embed.py`

File: mmseg/models/backbones/mit.py [hunk 17]
```diff
@@ -0,0 +1,454 @@
...
+                 drop_path_rate=0.,
+                 act_cfg=dict(type='GELU'),
+                 norm_cfg=dict(type='LN', eps=1e-6),
+                 sr_ratios=[8, 4, 2, 1],
+                 pretrain_style='official',
+                 pretrained=None,
+                 init_cfg=None):
+        super().__init__()
+
+        assert pretrain_style in ['official', 'mmcls']
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:353]: We may add an assert message.

File: mmseg/models/backbones/mit.py [hunk 18]
```diff
@@ -0,0 +1,454 @@
...
+    """
+
+    def __init__(self,
+                 in_channels=3,
+                 embed_dims=[64, 128, 256, 512],
+                 num_layers=[3, 4, 6, 3],
+                 num_heads=[1, 2, 4, 8],
+                 mlp_ratios=[4, 4, 4, 4],
+                 out_indices=(0, 1, 2, 3),
+                 qkv_bias=True,
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:341]: Add an assert for the length of related `args` (refer to [ResNet](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/backbones/resnet.py#L447)).

File: mmseg/models/backbones/mit.py [hunk 19]
```diff
@@ -0,0 +1,454 @@
...
+
+    A PyTorch implement of : `SegFormer: Simple and Efficient Design for
+    Semantic Segmentation with Transformers` -
+        https://arxiv.org/pdf/2105.15203.pdf
+    """
+
+    def __init__(self,
+                 in_channels=3,
+                 embed_dims=[64, 128, 256, 512],
+                 num_layers=[3, 4, 6, 3],
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:337]:
```suggestion
                 num_stages=4
                 stage_blocks=[3, 4, 6, 3],
```

File: mmseg/models/backbones/mit.py [hunk 20]
```diff
@@ -0,0 +1,454 @@
...
+
+        # transformer encoder
+        dpr = [
+            x.item()
+            for x in torch.linspace(0, drop_path_rate, sum(num_layers))
+        ]  # stochastic depth decay rule
+
+        cur = 0
+        self.layers = ModuleList()
+        for stage_id in range(len(num_layers)):
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:377]:
```suggestion
        for i, num_blocks in enumerate(self.stage_blocks):
```

Comment (Junjun2016) [reply mmseg/models/backbones/mit.py:377]: stage_id --> i

Comment (sennnnn) [reply mmseg/models/backbones/mit.py:377]: Stage_id -> id may cause conflict with TransformerEncoderLayer construction.

File: mmseg/models/backbones/mit.py [hunk 21]
```diff
@@ -0,0 +1,454 @@
...
+            num_heads=num_heads,
+            attn_drop=attn_drop_rate,
+            proj_drop=drop_rate,
+            dropout_layer=dict(type='DropPath', drop_prob=drop_path_rate),
+            batch_first=batch_first,
+            qkv_bias=qkv_bias,
+            norm_cfg=norm_cfg,
+            sr_ratio=sr_ratio)
+
+        _, self.norm2 = build_norm_layer(norm_cfg, embed_dims)
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:279]:
```suggestion
        self.norm2 = build_norm_layer(norm_cfg, embed_dims)[1]
```

File: mmseg/models/backbones/mit.py (hunk 12 above)
Comment (Junjun2016) [reply mmseg/models/backbones/mit.py:367]: We may also add padding.

File: mmseg/models/backbones/mit.py [hunk 22]
```diff
@@ -0,0 +1,454 @@
...
+                    drop_path_rate=dpr[cur + i],
+                    num_fcs=2,
+                    qkv_bias=qkv_bias,
+                    act_cfg=act_cfg,
+                    norm_cfg=norm_cfg,
+                    sr_ratio=sr_ratios[stage_id])
+                for i in range(num_layers[stage_id])
+            ])
+            in_channels = embed_dims[stage_id]
+            _, norm = build_norm_layer(norm_cfg, embed_dims[stage_id])
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:401]:
```suggestion
            norm = build_norm_layer(norm_cfg, embed_dims[stage_id])[1]
```

File: mmseg/models/backbones/mit.py [hunk 23]
```diff
@@ -0,0 +1,454 @@
...
+
+        self.ffn = MixFFN(
+            embed_dims=embed_dims,
+            feedforward_channels=feedforward_channels,
+            num_fcs=num_fcs,
+            ffn_drop=drop_rate,
+            dropout_layer=dict(type='DropPath', drop_prob=drop_path_rate),
+            act_cfg=act_cfg)
+
+    def forward(self, x, H, W):
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:289]: In MiT, the `NLC` shape is ONLY used in self-attention, so we can keep the shape of `NCHW` in the model and transform it twice in the transformer block. 
Transform `NCHW` to `NLC` before self-attention and transform `NLC` to `NCHW` after self-attention.

File: mmseg/models/backbones/mit.py [hunk 24]
```diff
@@ -0,0 +1,454 @@
...
+        _, _, H, W = x.shape
+        x = nchw_to_nlc(x)
+        x = self.norm(x)
+
+        return x, H, W
+
+
+@BACKBONES.register_module()
+class MixVisionTransformer(BaseModule):
+    """The backbone of Segformer.
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:327]: https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/backbones/resnet.py
https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/backbones/hrnet.py

File: mmseg/models/backbones/mit.py [hunk 25]
```diff
@@ -0,0 +1,434 @@
...
+from mmcv.cnn.bricks.transformer import MultiheadAttention
+from mmcv.runner import BaseModule, ModuleList, Sequential, _load_checkpoint
+
+from ...utils import get_root_logger
+from ..builder import BACKBONES
+from ..utils import PatchEmbed, mit_convert
+
+
+def nlc_to_nchw(tensor, H, W):
+    assert len(tensor.shape) == 3
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:19]: assert fo L and HW

File: mmseg/models/backbones/mit.py [hunk 26]
```diff
@@ -0,0 +1,434 @@
...
+    B, _, C = tensor.shape
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:20]:
```suggestion
    B, L, C = tensor.shape
```
and assert fo L and HW

File: mmseg/models/backbones/mit.py [hunk 27]
```diff
@@ -0,0 +1,434 @@
...
+def nchw_to_nlc(tensor):
+    assert len(tensor.shape) == 4
+    return tensor.flatten(2).transpose(1, 2).contiguous()
+
+
+class PEConv(BaseModule):
+    """Mix-FFN use 3x3 depth-wise Conv to provide positional encode
+    information.
+
+    Args:
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:33]: The complete docstring for `Args`.

File: mmseg/models/backbones/mit.py [hunk 28]
```diff
@@ -0,0 +1,434 @@
...
+            Default: None.
+    """
+
+    def __init__(self,
+                 embed_dims,
+                 feedforward_channels,
+                 num_fcs=2,
+                 act_cfg=dict(type='GELU'),
+                 ffn_drop=0.,
+                 pe_index=0.,
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:93]: The complete docstring for `Args`.

File: mmseg/models/backbones/mit.py [hunk 29]
```diff
@@ -0,0 +1,434 @@
...
+                 dropout_layer=None,
+                 add_identity=True,
+                 init_cfg=None):
+        super(MixFFN, self).__init__(init_cfg)
+        assert num_fcs >= 2, 'num_fcs should be no less ' \
+            f'than 2. got {num_fcs}.'
+
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:100]: assert for `pe_index`.

File: mmseg/models/backbones/mit.py [hunk 30]
```diff
@@ -0,0 +1,434 @@
...
+        conv1x1 = partial(
+            ConvModule,
+            kernel_size=1,
+            stride=1,
+            bias=True,
+            norm_cfg=None,
+            act_cfg=None)
+
+        layers = []
+        in_channels = embed_dims
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:116]: `pe_index` belongs to [0, num_fcs].
`pe_index=0` --> before the first `1x1 conv` or `FC`.
`pe_index=num_fcs` --> after the last `1x1 conv` or `FC`.

File: mmseg/models/backbones/mit.py [hunk 31]
```diff
@@ -0,0 +1,434 @@
...
+
+        self.sr_ratio = sr_ratio
+        if sr_ratio > 1:
+            self.sr = ConvModule(
+                in_channels=embed_dims,
+                out_channels=embed_dims,
+                kernel_size=sr_ratio,
+                stride=sr_ratio,
+                norm_cfg=None,
+                act_cfg=None)
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:204]: It would be preferable if these were args.

File: mmseg/models/backbones/mit.py [hunk 32]
```diff
@@ -0,0 +1,434 @@
...
+        assert pretrain_style in [
+            'official', 'mmcls'
+        ], 'we only support official weights or mmcls weights.'
+
+        if isinstance(pretrained, str) or pretrained is None:
+            warnings.warn('DeprecationWarning: pretrained is a deprecated, '
+                          'please use "init_cfg" instead')
+        else:
+            raise TypeError('pretrained must be a str or None')
+
```
Comment (Junjun2016) [inline mmseg/models/backbones/mit.py:341]: Adding assert for the length of some `Args` would be better.

File: mmseg/models/utils/embed.py [hunk 33]
```diff
@@ -42,6 +42,8 @@ def __init__(self,
         if stride is None:
             stride = kernel_size
 
+        self.overlapping = stride == kernel_size
+
         # The default setting of patch size is eaual to kernel size.
```
Comment (Junjun2016) [inline mmseg/models/utils/embed.py:47]:
```suggestion
        # The default setting of patch size is equal to kernel size.
```

File: mmseg/models/utils/embed.py [hunk 34]
```diff
@@ -73,12 +75,18 @@ def __init__(self,
...
-            x = F.pad(x,
-                      (0, 0, 0, self.patch_size[0] - H % self.patch_size[0]))
-        if W % self.patch_size[1] != 0:
-            x = F.pad(x,
-                      (0, self.patch_size[1] - W % self.patch_size[1], 0, 0))
+
+        # Modify H, W to multiple of patch size.
+        if self.overlapping:
+            pass
+        else:
```
Comment (Junjun2016) [inline mmseg/models/utils/embed.py:82]:
```suggestion
        if  not self.overlapping:
```

File: mmseg/models/backbones/mit.py (hunk 31 above)
Comment (sennnnn) [reply mmseg/models/backbones/mit.py:204]: act_cfg control all of the activation layer build. norm_cfg control all of the norm layer build. Add more prefix or suffix may be confused.

File: mmseg/models/utils/embed.py (hunk 34 above)
Comment (sennnnn) [reply mmseg/models/utils/embed.py:82]: It's just a todo placeholder.

File: mmseg/models/utils/shape_convert.py [hunk 35]
```diff
@@ -0,0 +1,10 @@
+def nlc_to_nchw(tensor, H, W):
```
Comment (xvjiarui) [inline mmseg/models/utils/shape_convert.py:1]: Docstring.

File: mmseg/models/backbones/mit.py [hunk 36]
```diff
@@ -0,0 +1,479 @@
...
+    norm_cfg (dict): Config dict for normalization layer.
+        Default: dict(type='LN')
+    act_cfg (dict): The activation config for FFNs.
+        Defalut: dict(type='GELU').
+    pretrain_style (str): Choose to use official or mmcls pretrain weights.
+        Default: official.
+    pretrained (str, optional): model pretrained path. Default: None.
+    init_cfg (dict or list[dict], optional): Initialization config dict.
+        Default: None.
+    """
```
Comment (xvjiarui) [inline mmseg/models/backbones/mit.py:343]: The format of the docstring is not correct.

File: mmseg/models/backbones/mit.py [hunk 37]
```diff
@@ -0,0 +1,479 @@
...
+
+        self.ffn = MixFFN(
+            embed_dims=embed_dims,
+            feedforward_channels=feedforward_channels,
+            num_fcs=num_fcs,
+            ffn_drop=drop_rate,
+            dropout_layer=dict(type='DropPath', drop_prob=drop_path_rate),
+            act_cfg=act_cfg)
+
+    def forward(self, x, H, W):
```
Comment (xvjiarui) [inline mmseg/models/backbones/mit.py:295]:
```suggestion
    def forward(self, x, hw_shape):
```

File: mmseg/models/backbones/mit.py [hunk 38]
```diff
@@ -0,0 +1,479 @@
...
+        self.act_cfg = act_cfg
+        self.activate = build_activation_layer(act_cfg)
+
+        conv1x1 = partial(
+            ConvModule,
+            kernel_size=1,
+            stride=1,
+            bias=True,
+            norm_cfg=None,
+            act_cfg=None)
```
Comment (xvjiarui) [inline mmseg/models/backbones/mit.py:105]: No need to use `partial`.

File: mmseg/models/backbones/mit.py [hunk 39]
```diff
@@ -0,0 +1,479 @@
...
+from mmcv.cnn.bricks.drop import build_dropout
+from mmcv.cnn.bricks.transformer import MultiheadAttention
+from mmcv.runner import BaseModule, ModuleList, Sequential, _load_checkpoint
+
+from ...utils import get_root_logger
+from ..builder import BACKBONES
+from ..utils import PatchEmbed, mit_convert, nchw_to_nlc, nlc_to_nchw
+
+
+class PEConv(BaseModule):
```
Comment (xvjiarui) [inline mmseg/models/backbones/mit.py:18]: Why we need to use `PEConv` to wrap `conv`?

File: mmseg/models/backbones/mit.py [hunk 40]
```diff
@@ -0,0 +1,479 @@
...
+
+    def __init__(self,
+                 embed_dims,
+                 feedforward_channels,
+                 num_fcs=2,
+                 act_cfg=dict(type='GELU'),
+                 ffn_drop=0.,
+                 pe_index=1,
+                 dropout_layer=None,
+                 add_identity=True,
```
Comment (xvjiarui) [inline mmseg/models/backbones/mit.py:84]: Is this argument necessary?

File: mmseg/models/backbones/mit.py [hunk 41]
```diff
@@ -0,0 +1,479 @@
...
+            conv1x1(
+                in_channels=feedforward_channels, out_channels=in_channels))
+        # Last position of MixFFN
+        if pe_index == num_fcs:
+            layers.append(PEConv(feedforward_channels))
+        layers.append(nn.Dropout(ffn_drop))
+        self.layers = Sequential(*layers)
+        self.dropout_layer = build_dropout(
+            dropout_layer) if dropout_layer else torch.nn.Identity()
+        self.add_identity = add_identity
```
Comment (xvjiarui) [inline mmseg/models/backbones/mit.py:134]: This part is too long. 
Consider simplifying it.

File: mmseg/models/backbones/mit.py [hunk 42]
```diff
@@ -0,0 +1,479 @@
...
+            Default: None.
+    """
+
+    def __init__(self,
+                 embed_dims,
+                 feedforward_channels,
+                 num_fcs=2,
+                 act_cfg=dict(type='GELU'),
+                 ffn_drop=0.,
+                 pe_index=1,
```
Comment (xvjiarui) [inline mmseg/models/backbones/mit.py:82]: If it is too complicated, we may remove `pe_index`.

Comment (xvjiarui) [reply mmseg/models/backbones/mit.py:82]: Just fix the`num_fcs` to 2, and insert `PE` in the middle.

File: mmseg/models/utils/shape_convert.py [hunk 43]
```diff
@@ -0,0 +1,22 @@
+def nlc_to_nchw(x, hw_shape):
+    """Convert [N, L, C] shape tensor to [N, C, H, W] shape tensor.
+
+    Args:
+        x (Tensor): The input tensor for convertion.
+        hw_shape (Sequence[int]): The height and width of output feature map.
+    """
```
Comment (xvjiarui) [inline mmseg/models/utils/shape_convert.py:7]: Return

File: mmseg/models/utils/shape_convert.py [hunk 44]
```diff
@@ -0,0 +1,22 @@
...
+    B, L, C = x.shape
+    assert L == H * W, 'The seq_len doesn\'t match H, W'
+    return x.transpose(1, 2).reshape(B, C, H, W)
+
+
+def nchw_to_nlc(x):
+    """Flatten [N, C, H, W] shape tensor to [N, L, C] shape tensor.
+
+    Args:
+        x (Tensor): The input tensor for convertion.
```
Comment (xvjiarui) [inline mmseg/models/utils/shape_convert.py:19]: Return

File: mmseg/models/utils/embed.py [hunk 45]
```diff
@@ -73,12 +75,17 @@ def __init__(self,
...
         H, W = x.shape[2], x.shape[3]
-        if H % self.patch_size[0] != 0:
-            x = F.pad(x,
-                      (0, 0, 0, self.patch_size[0] - H % self.patch_size[0]))
-        if W % self.patch_size[1] != 0:
-            x = F.pad(x,
-                      (0, self.patch_size[1] - W % self.patch_size[1], 0, 0))
+
+        # TODO: Process overlapping op
+        if not self.overlapping:
```
Comment (xvjiarui) [inline mmseg/models/utils/embed.py:80]: Overlapping is not precise for this. 
We may consider something like `auto_pad` or `pad_to_patch_size`. 

We also need to make it an argument and add a docstring for that.

File: mmseg/models/utils/embed.py [hunk 46]
```diff
@@ -56,7 +58,7 @@ def __init__(self,
         self.patch_size = patch_size
 
         # Use conv layer to embed
-        conv_type = conv_type or dict(type='Conv2d')
+        conv_type = conv_type or 'Conv2d'
```
Comment (xvjiarui) [inline mmseg/models/utils/embed.py:61]: Could we use `conv_cfg`?

File: mmseg/models/utils/embed.py [hunk 47]
```diff
@@ -19,6 +19,8 @@ class PatchEmbed(BaseModule):
             Default: None (Default to be equal with kernel_size).
         padding (int): The padding length of embedding conv. Default: 0.
         dilation (int): The dilation rate of embedding conv. Default: 1.
+        pad_to_patch_size (bool, optional): Whether to pad feature map shape
+            to multiple patch size. Default: False.
```
Comment (xvjiarui) [inline mmseg/models/utils/embed.py:23]: We may make it `True` by default.

**Review by xvjiarui — APPROVED** [review]

**Pull request merged**